### PR TITLE
disable recompaction

### PR DESCRIPTION
**Pull request by iilyak**
In some cases it is a beneficial to be able to temporary disable recompaction. 
This PR introduces a way to do so globally or individually per database or specific index.

The configuration of the feature is as follows:

```
[view_compaction.recompaction]
enable = true / false %% to set globaly
<db_name> = enable / disable %% to set per db
<db_name>:<idx_name> = enable / disable %% to set per index
```

Depends on: https://github.com/apache/couchdb-erlang-tests/pull/2

### CODE REVIEW

**Comment by rnewson** [comment]
compaction in couchdb is initiated by the user or the compaction daemon, this seems to be in the wrong place.

**Comment by kxepal** [comment]
Recompaction? What does it gives?

**Comment by iilyak** [comment]
@rnewson: This change is for view_compaction (not db compaction). Here is my understanding of how it would work. 

- The entry point to trigger compaction is [`couch_index:compact`](https://github.com/cloudant/couchdb-couch-index/blob/a3af68f86ee21a9e072507b9d1d7da11e001b299/src/couch_index.erl#L76). 
- It will do [gen_server:call(CPid, compact)](https://github.com/apache/couchdb-couch-index/blob/master/src/couch_index.erl#L77) to compactor Pid. 
- `couch_index_compactor` would [spawn_link(fun() -> compact/3..](https://github.com/apache/couchdb-couch-index/blob/master/src/couch_index_compactor.erl#L64). 
- `couch_index_compactor:compact/3` would call [Mod:compact/3](https://github.com/apache/couchdb-couch-index/blob/master/src/couch_index_compactor.erl#L104). 
In our case `Mod` would be `couch_mrview_index`.
- `couch_mrview_index:compact/3` would dispatch the call to [couch_mrview_compactor:compact/3](https://github.com/apache/couchdb-couch-mrview/blob/master/src/couch_mrview_index.erl#L193). 
- When compaction is finished we get control back into [couch_index_compactor:compact](https://github.com/apache/couchdb-couch-index/blob/master/src/couch_index_compactor.erl#L106). 
- Where right after `couch_mrview_compactor` we do `Mod:commit(NewIdxState)`. We notify couch_index by doing [gen_server:call(Idx, {compacted, NewIdxState})](https://github.com/apache/couchdb-couch-index/blob/master/src/couch_index_compactor.erl#L107). 
- I.e. we get [here](https://github.com/cloudant/couchdb-couch-index/blob/a3af68f86ee21a9e072507b9d1d7da11e001b299/src/couch_index.erl#L191). Where we can check if re-compaction is enabled. 
- If re-compaction is not enabled we swap files. 
Otherwise we reply `{reply, recompact, State}`. So  we can trigger [couch_mrview_index:compact(_, _, [recompact])](https://github.com/cloudant/couchdb-couch-index/blob/a3af68f86ee21a9e072507b9d1d7da11e001b299/src/couch_index_compactor.erl#L108:L110).

Please let me know what I'm missing.

**Comment by iilyak** [comment]
@kxepal: If recompaction is enabled it would through away "almost" compacted view if db update occurs. Sometimes though you want to let compaction finish and then do regular index update to catch up. By temporary disabling recompaction you would be able to do so. Above is just one example when it could be useful. There are other use cases as well.

File: src/couch_index.erl [hunk 1]
```diff
@@ -407,3 +394,160 @@ assert_signature_match(Mod, OldIdxState, NewIdxState) ->
...
+    State#st{
+        idx_state=NewIdxState1,
+        committed=false
+     }.
+
+is_recompaction_enabled(IdxState, #st{mod = Mod}) ->
+    DbName = binary_to_list(Mod:get(db_name, IdxState)),
+    IdxName = binary_to_list(Mod:get(idx_name, IdxState)),
+    IdxKey = DbName ++ ":" ++ IdxName,
+    Global = config:get("view_compaction.recompaction", "enable"),
```
Comment (davisp) [inline src/couch_index.erl]: I think you should be using "enabled" and "disabled" here. The noun version I think is the more common use for a setting value.

File: src/couch_index.erl [hunk 2]
```diff
@@ -407,3 +394,160 @@ assert_signature_match(Mod, OldIdxState, NewIdxState) ->
...
+    PerIndex = config:get("view_compaction.recompaction", IdxKey),
+    PerDb = config:get("view_compaction.recompaction", DbName),
+    case {Global, PerDb, PerIndex} of
```
Comment (davisp) [inline src/couch_index.erl]: This truth table is confusing to me. I'd disable to on and then have a trigger off with a simple orelse combo. Something along the lines of:

    Global = config:get("view_compaction.recompaction", "enabled", "true") == "false",
    PerDb = config:get("view_compaction.recompaction", DbName, "enabled") == "disabled",
    PerIdx = config:get("view_compaction.recompaction", IdxKey, "enabled") == "disabled",
    Global orelse PerDb orelse PerIdx.

File: src/couch_index.erl (hunk 1 above)
Comment (davisp) [inline src/couch_index.erl]: Also true vs enabled for settings seems odd. Seems like you'd be better off using something like "status = enabled" as a setting. Or putting it in a different section for the global settings although I dunno if we have a better section. Though you could just use

    [view_compaction]
    recompaction = enabled

    [view_compaction.recompaction]
    DbName = enabled | disabled
    DbName:IdxName = enabled | disabled

Comment (kxepal) [inline src/couch_index.erl]: I would prefer
```
[view_compaction]
enable_recompaction = true | false

[view_compaction.DbName]
enable = true | false
[view_compaction.DbName:IdxName]
enable = true | false
```

Because:
- This is boolean logic without else state;
- For others switches we use booleans as well;
- If we would like to add few more options for recompaction (still don't understand what is it and why), there wouldn't be any issues with;
- It's more error-prone solution (enable or enabled?)

**Comment by davisp** [comment]
For background on this, "recompaction" is just an internal name for the step during view compaction when we spawn an index updater to top off a compacted view. This has been in CouchDB since almost forever but was never used by Cloudant.

The reason we have the view updater is that (unlike DB compactions) we can't resume a view compaction. So we have to make sure that when we swap out the compaction that there's no need to retry compaction.

The downside to not running the recompact step is that a view will appear to roll backwards when a view compaction finishes. In a cluster this isn't a big deal because there are two other copies that will cover while the index updater catches back up.

The downside to running a recompact is that if the update frequency on a database is too high we'll never converge on view compactions. Which makes obvious sense if you think about it. Any updates that happen during the initial compaction stage need to be re-processed by the index updater. If you have two index updaters running over the same data (ie, one updating the uncompacted index, one trying to finish updates on the compacted index) they should be running at the same rate and will thus never converge.

To @rnewson's comment, this is the appropriate place. Its not a question of whether/when compaction runs, but whether we perform the index updater step which we have long known falls down on large views and/or high update rates.

For the config layout, I'd propose a combined approach between mine and @kxepal's suggestions. I do like the booleans but I think the `[view_compaction.recompact] with `$dbname` and `$dbname:$view` is more appropriate. The specific sections with a single pre-defined key seem like a misuse of the `section/key/value` namespace config hierarchy.

**Comment by kxepal** [comment]
@davisp 
Thanks you for the great (as usual) explanation! +1 for combined config approach, also looks good.

File: src/couch_index.erl (hunk 2 above)
Comment (iilyak) [inline src/couch_index.erl]: It is about defaults. By default recompaction should be enabled. We also need to distinguish between default and user provided value. There are two major modes:

1. when `view_compaction.enable_recompaction = true`
   In this case we can set `view_recompaction.DbName:IdxName.enable = false` to ***disable*** `recompact`.

1. when `view_compaction.enable_recompaction = false`
   In this case we can set `view_recompaction.DbName:IdxName.enable = true` to ***enable*** `recompact`.

Although I agree the case statement is hard to follow. I'll simplify the condition.

**Comment by iilyak** [comment]
@davisp: The hard to follow case statement was removed. I'll fix the commit message to document the new names of the settings when squash.

File: src/couch_index.erl (hunk 2 above)
Comment (davisp) [inline src/couch_index.erl]: Fair enough. I didn't contemplate an operator wanting to selectively re-enable recompact at a narrower scope when its been disabled in a wider scope. I guess that could be a possibility.

**Comment by davisp** [comment]
@iilyak +1
